_posts/2022-06-08-packaging-functions-with-helm.md

```diff
@@ -0,0 +1,261 @@
+
+![Using the CRD to deploy functions](/images/2021-06-kubectl-functions/operator-crd.png)
+> Using the CRD to deploy functions
+
+## Deploy functions using the Function CRD
+We will look into two different approaches to deploy functions when using OpenFaaS with the Function Custom Resource Definition and Operator mode. The first method will use the `faas-cli` to convert the OpenFaaS stack.yaml file into a Kubernetes CRD YAML file. The resulting files can be deployed using `kubectl` or committed to a repo where GitOps tools like [ArgoCD](https://argo-cd.readthedocs.io/en/stable/) or [Flux](https://fluxcd.io/) could use them to deploy the functions. In the second approach the functions are packaged as a Helm Chart. Helm can then be used to deploy the functions and provide configuration values to the deployment. Both are valid options and they come with their own benefits and drawbacks.
+
+CRD files:
+- CRDs can be generated from a stack.yaml file using `faas-cli generate`
+- Basic templating support using [YAML environment variable substitution](https://docs.openfaas.com/reference/yaml/#yaml-environment-variable-substitution)
+
+Helm Chart:
+- Once templates are generated you can not just run `faas-cli generate` again to reflect changes in your stack.yaml file.
+- Because Helm uses the [Go template language](https://pkg.go.dev/text/template) it supports advanced templating (logic, functions).
+- Values are centralised in the values.yaml file.
+- Charts can be versioned and Helm provides mechanisms to easily upgrade, rollback and uninstall said Charts.
+- Helm provides you the option to group various Kubernetes components together with your functions into a single deployable unit.
+
+Using Helm to package functions does not mean you wont need the YAML environment variable substitution anymore. While the Helm templates contain the deployment information, the stack.yaml file still contains the build configuration for functions. Environment variable substitution can be used to provide different configurations for building functions locally vs building and pushing functions in a CI environment.
+
+All fields that are available to configure function deployments in the stack.yaml file can also be used in a Function CRD.
```

Comment:
Any configuration about building the function is left out, such as the template name and build args.